File: 프로그래머스/lv2/42883. 큰 수 만들기/큰 수 만들기.js
// 앞에서부터 숫자를 스택에 push
// push 하려는 값이 스택 마지막 값보다 큰지 비교한다.
// 넣으려는 값이 크면 스택에서 pop을 한다.
// 위 과정 반복

function solution(number, k) {
    const stack = [];
    let cnt = k;
    
    for(const n of number){
        while(cnt > 0 && stack[stack.length-1] < n){
            stack.pop();
            cnt-=1;
        }
        stack.push(n);
    }
    
    for(let i=0; i<cnt; i++) stack.pop();
    
    return stack.join('');

}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    stack = []
    count = 0

    for n in number:
        
        while len(stack)!=0 and count<k and n > stack[-1]:
            stack.pop()
            count+=1
        
        stack.append(n)
    
    # count가 남았을 때! 뒤에서 제거해주기
    for i in range(k-count):
        stack.pop()
    
    return "".join(stack)